### PR TITLE
rust scheduler UI changes

### LINKED ISSUE #18: `ScheduleToken::erase` is unused, remove?
We have this definition in `crates/bindings/src/lib.rs`:
```rust
    /// Erase the `R` type parameter from the token.
    ///
    /// In other words, forget what reducer this is for.
    #[inline]
    pub fn erase(self) -> ScheduleToken {
        ScheduleToken::new(self.id)
    }
```

It seems unused so it's unclear if it should be provided.

### PR DESCRIPTION
**Pull request by Shubham8287**
# Description of Changes

API/ABI chages for scheduled table as described - https://github.com/clockworklabs/SpacetimeDBPrivate/blob/master/proposals/0005-timers.md#api-design 

Closes #18.

# API and ABI breaking changes
1

### CODE REVIEW

File: crates/bindings-macro/src/lib.rs [hunk 1]
```diff
@@ -436,13 +457,61 @@ struct Column<'a> {
...
+// add scheduled_id and scheduled_at fields to the struct
+fn add_scheduled_fields(item: &mut DeriveInput) -> syn::Result<()> {
+    match &mut item.data {
+        syn::Data::Struct(ref mut struct_data) => match &mut struct_data.fields {
+            syn::Fields::Named(fields) => {
+                fields.named.extend([
+                    syn::Field::parse_named.parse2(quote! {#[autoinc]
+                    pub scheduled_id: u64 })?,
+                    syn::Field::parse_named
+                        .parse2(quote! { pub scheduled_at: spacetimedb::spacetimedb_lib::ScheduleAt })?,
```
Comment (RReverser) [inline crates/bindings-macro/src/lib.rs:494]: In the spec this is called `at`, but here it's called `scheduled_at` - which one should be updated?

File: crates/bindings-macro/src/lib.rs [hunk 2]
```diff
@@ -436,13 +457,61 @@ struct Column<'a> {
...
+}
+
+// add scheduled_id and scheduled_at fields to the struct
+fn add_scheduled_fields(item: &mut DeriveInput) -> syn::Result<()> {
+    match &mut item.data {
+        syn::Data::Struct(ref mut struct_data) => match &mut struct_data.fields {
+            syn::Fields::Named(fields) => {
+                fields.named.extend([
+                    syn::Field::parse_named.parse2(quote! {#[autoinc]
+                    pub scheduled_id: u64 })?,
```
Comment (RReverser) [inline crates/bindings-macro/src/lib.rs:492]: In the spec this is called `id`, but here it's called `scheduled_id` - which one should be updated?

File: crates/bindings-macro/src/lib.rs (hunk 1 above)
Comment (Shubham8287) [reply crates/bindings-macro/src/lib.rs:494]: Let's refer them as `scheduled_id` and `scheduled_at`, I will update same in proposal.

File: crates/bindings-macro/src/lib.rs [hunk 3]
```diff
@@ -436,13 +457,61 @@ struct Column<'a> {
...
+    Ok(output)
+}
+
+// add scheduled_id and scheduled_at fields to the struct
+fn add_scheduled_fields(item: &mut DeriveInput) -> syn::Result<()> {
+    match &mut item.data {
+        syn::Data::Struct(ref mut struct_data) => match &mut struct_data.fields {
+            syn::Fields::Named(fields) => {
+                fields.named.extend([
+                    syn::Field::parse_named.parse2(quote! {#[autoinc]
```
Comment (RReverser) [inline crates/bindings-macro/src/lib.rs:491]: The spec says

> and an auto-increment primary key

but this one is just `[autoinc]`.

Comment (Shubham8287) [reply crates/bindings-macro/src/lib.rs:491]: After implementation I realized that this does not need to be a primary key, just `[autoinc]` is fine.

Comment (RReverser) [reply crates/bindings-macro/src/lib.rs:491]: I think primary key is still better for indexing, no?

File: crates/sdk/tests/test.rs [hunk 4]
```diff
@@ -146,4 +146,4 @@ macro_rules! declare_tests_with_suffix {
 }
 
 declare_tests_with_suffix!(rust, "");
-declare_tests_with_suffix!(csharp, "-cs");
+//declare_tests_with_suffix!(csharp, "-cs");
```
Comment (Shubham8287) [inline crates/sdk/tests/test.rs:149]: Address this before merging

File: crates/bindings-macro/src/lib.rs [hunk 5]
```diff
@@ -721,10 +776,19 @@ fn spacetimedb_tabletype_impl(item: syn::DeriveInput) -> syn::Result<TokenStream
             Span::call_site(),
         )
     });
+
+    let scheduled_constant = match sats_ty.scheduled {
+        Some(reducer_name) => quote!(const SCHEDULED_REDUCER_NAME: Option<&'static str> = Some(#reducer_name);),
+        None => quote!(
+            const SCHEDULED_REDUCER_NAME: Option<&'static str> = None;
+        ),
+    };
```
[the remark this answers is not in the document]
Comment (Shubham8287) [reply crates/bindings-macro/src/lib.rs:785]: This was having an isuee, If I remember correctly.
`quote!` just ignore `None`, so this was rendering as
```
const SCHEDULED_REDUCER_NAME: Option<&'static str> = ;
```

I will check once again, there is a chance I might have been doing something different.

File: crates/sats/src/db/def.rs [hunk 6]
```diff
@@ -1053,6 +1053,13 @@ impl TableDef {
         x
     }
 
+    /// Set the reducer name for scheduled tables and return updated `TableDef`.
+    pub fn with_scheduled(self, scheduled: Option<Box<str>>) -> Self {
+        let mut x = self;
+        x.scheduled = scheduled;
+        x
```
[the remark this answers is not in the document]
Comment (Shubham8287) [reply crates/sats/src/db/def.rs:1060]: lol, I also found the code here wierd but just followed pattern from above.

File: modules/rust-wasm-test/src/lib.rs [hunk 7]
```diff
@@ -80,9 +80,18 @@ pub type TestAlias = TestA;
 // #[spacetimedb(migrate)]
 // pub fn migrate() {}
 
+#[spacetimedb(table, scheduled(repeating_test))]
```
[the remark this answers is not in the document]
Comment (Shubham8287) [reply modules/rust-wasm-test/src/lib.rs:83]: just checked, it does not!

File: crates/bindings/src/timestamp.rs [hunk 8]
```diff
@@ -81,6 +81,12 @@ impl Timestamp {
     }
 }
 
+impl From<Timestamp> for spacetimedb_lib::Timestamp {
+    fn from(ts: Timestamp) -> Self {
+        Self(ts.into_micros_since_epoch())
+    }
+}
```
[the remark this answers is not in the document]
Comment (Shubham8287) [reply crates/bindings/src/timestamp.rs:88]: that may not be super straightforward, we may get cyclic imports.

File: crates/bindings-macro/src/lib.rs [hunk 9]
```diff
@@ -189,7 +195,21 @@ impl syn::parse::Parse for MacroInput {
...
-                Self::Table(public)
+
+                let mut scheduled = None;
+                comma_then_comma_delimited(input, || {
+                    match_tok!(match input {
+                        kw::scheduled => {
+                            let in_parens;
+                            syn::parenthesized!(in_parens in input);
+                            let in_parens = &in_parens;
+                            scheduled = Some(in_parens.parse::<Ident>()?);
```
[the remark this answers is not in the document]
Comment (Shubham8287) [reply crates/bindings-macro/src/lib.rs:209]: Fixed this, leaving comment on you to verify and resolve.

File: modules/rust-wasm-test/src/lib.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (Shubham8287) [reply modules/rust-wasm-test/src/lib.rs:83]: fixed this

File: crates/bindings/src/timestamp.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (Shubham8287) [reply crates/bindings/src/timestamp.rs:88]: I mean moving `TimeStamp` to primitives will also requires importing from `_sats` there, hence cyclic import. Isn't?

**Pull request merged**